### ISSUE TITLE
log facility overwrite

### ISSUE THREAD

**Issue by pubyun** [haproxy/haproxy#1901]
### Detailed Description of the Problem

here is a blog:

https://www.haproxy.com/blog/log-forwarding-with-haproxy-and-syslog/

```
log-forward syslog
  # Accepts incoming TCP messages
  bind 0.0.0.0:514

  # Accepts incoming UDP messages
  dgram-bind 0.0.0.0:514

  # Sends outgoing messages via UDP
  log 172.25.0.12:514 local0

```

the blog said:
```
We’re listening and sending using port 514, which is the standard Syslog port for both UDP and TCP. The facility code, local0, is used only if the received log event doesn’t already set a facility.
```

i send a message to haproxy:

logger -p daemon.warn "This is only a test."

and the logger received a message with facility local0.

### Expected Behavior

haproxy only use local0 if the received log event doesn’t already set a facility.

### Steps to Reproduce the Behavior

logger -p daemon.warn "This is only a test."

### What is your configuration?

```haproxy
provided
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.19-1 2022/09/30 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.19.html
Running on: Linux 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64
```

**Comment by Darlelet** [comment]
Hi,

Did you try the logger command without haproxy running?
I suspect that logger command is directly speaking to rsyslogd, thus bypassing your haproxy log-forward.

Please try using a different port for haproxy forwarder so we know for sure that it is not related:
`logger -d -s 127.0.0.1 -P custom_port`
(adjust your bind and dgram-bind directives accordingly)

-----------------------

Moreover, logger doesn't seem like the right tool to test this feature.

It is true that logger -p command allows you to provide "facility.severity" hint.

But according to logger documentation:
```
   -p, --priority priority
           Enter the message into the log with the specified priority. The priority may be specified numerically or as a facility.level pair. For example, -p local3.info logs the message as informational in the local3 facility. The
           default is user.notice
```
It always falls back to **user** facility if you don't explicitely provide one.
Thus, logger will never send a syslog message without the "facility" set to >= user:
This results in haproxy systematically ignoring the facility parameter in your log-forward configuration.

Instead you could test the feature using manually crafted incomplete syslog message (missing the facility).

Here is a working example:

```
log-forward syslog
  # Accepts incoming UDP messages
  dgram-bind 0.0.0.0:9999

  # Sends outgoing messages via UDP
  log 127.0.0.1:514 format rfc5424 user
```
(haproxy.conf)

Note that I added "format rfc5424" because according to haproxy doc, facility parameter is ignored if facility is not mandatory for current log format:

> If no format specified, HAProxy tries to keep the incoming log format.
  Configured facility is ignored, except if incoming message does not
  present a facility but one is mandatory on the outgoing format.
  If there is no timestamp available in the input format, but the field
  exists in output format, HAProxy will use the local date.


Sending a manually crafted payload (without the facility) to 127.0.0.1:9999 (haproxy log-forwarder):
```
echo -ne 'test without facility' | nc -u 127.0.0.1 9999 -w0
```

The corresponding message forwarded by Haproxy to 127.0.0.1:514:
```
<8>1 2022-10-18T11:54:55.414751+02:00 - - - - - test without facility
```

Facility has automatically been set to "user" by HAProxy before forwarding the message.

If I change 
`log 127.0.0.1:514 format rfc5424 user`
to
`log 127.0.0.1:514 format rfc5424 local0`
in haproxy conf

The corresponding message forwarded by Haproxy on 127.0.0.1:514 will be:
```
<128>1 2022-10-18T11:55:16.881422+02:00 - - - - - test without facility
```

**Comment by pubyun** [comment]
sorry, i copy the config from blog and not my running config. here is my running config, the tcp forwarder  will overwrite log facility:

```
log-forward syslog
  bind xxx.xxx.169.49:514
  dgram-bind 118.184.169.49:514
  log ring@logbuffer local5

ring logbuffer
  description "buffer for logs"
  format rfc5424
  maxlen 1200
  size 32764
  timeout connect 5s
  timeout server 10s

  # Sends outgoing messages via TCP
  server logserver 172.18.11.33:5514 check log-proto octet-count
```

if i separate udp and tcp binding, the tcp forwarder will OK, the log facility is not overwrited:

```
log-forward syslog-udp
  dgram-bind  xxx.xxx.169.49:514
  log 172.18.11.33:5514 local3

log-forward syslog-tcp
  bind  xxx.xxx.169.49:514
  log ring@logbuffer local5

ring logbuffer
  description "buffer for logs"
  format rfc5424
  maxlen 1200
  size 32764
  timeout connect 5s
  timeout server 10s

  # Sends outgoing messages via TCP
  server logserver 172.18.11.33:5514 check log-proto octet-count
```

**Comment by capflam** [comment]
There is indeed a bug. In fact, it is not related to the bind lines but to the log target. If a server is configured, the facility from the original message is kept. However, if a ring is specified, the configured facility is always used.
Here is the fix:

```diff
diff --git a/src/log.c b/src/log.c
index 07b6b75e45..b3865e99e4 100644
--- a/src/log.c
+++ b/src/log.c
@@ -1681,13 +1681,13 @@ static inline void __do_send_log(struct logsrv *logsrv, int nblogger, int level,
        if (logsrv->type == LOG_TARGET_BUFFER) {
                struct ist msg;
 
                msg = ist2(message, size);
                msg = isttrim(msg, logsrv->maxlen);
 
-               sent = sink_write(logsrv->sink, &msg, 1, level, logsrv->facility, metadata);
+               sent = sink_write(logsrv->sink, &msg, 1, level, facility, metadata);
        }
        else if (logsrv->addr.ss_family == AF_CUST_EXISTING_FD) {
                struct ist msg;
 
                msg = ist2(message, size);
                msg = isttrim(msg, logsrv->maxlen);
```

I'll push the fix. Thanks !

**Comment by pubyun** [comment]
fixed, thanks

**Comment by capflam** [comment]
Now fixed in 2.7-dev. The patch will be backported soon. Thanks !
